### PR TITLE
let scylla-gdb.py recognize coroutines

### PR DESCRIPTION
**Pull request by cmm**
"scylla task_histogram" and "scylla fiber" will now show coroutine "promises".

Refs #10894

### CODE REVIEW

File: scylla-gdb.py [hunk 1]
```diff
@@ -1295,6 +1296,11 @@ def _make_symbol_matchers(symbol_specs):
     def _make_symbol_matcher(symbol_spec):
         unmangled_prefix = 'vtable for {}'.format('::'.join(symbol_spec))
         def matches_symbol(name):
+            name = name.strip()
+            if name.endswith(")") or (") [clone ." in name and name.endswith("]")):
```
Comment (psarna) [inline scylla-gdb.py:1300]: `endswith(")")` indeed sounds rather generic; did you test manually how many false positives are actually found when calling `scylla task_histogram` or any other function that uses this code?

Comment (denesb) [inline scylla-gdb.py:1300]: Why is a name ending with `)` enough to qualify it as a coroutine? What suffix are you are looking for? The commit message suggests `[clone .resume]`, where does the `)` come in here?

**Comment by denesb** [comment]
We have this issue: https://github.com/scylladb/scylla/issues/10894
@psarna, @piodul do you consider this PR to fix the above issue? If not, what else is missing?

File: scylla-gdb.py (hunk 1 above)
Comment (cmm) [reply scylla-gdb.py:1300]: tried it on the nice huge coredump from #10933, didn't see any false positives actually, but who knows

**Comment by psarna** [comment]
I think it's a decent start, so we can mark this pull request as `Fixes #10894`

File: scylla-gdb.py (hunk 1 above)
Comment (cmm) [reply scylla-gdb.py:1300]: `[clone .resume]` is one possibility (those probably correspond to `co_await` points in the middle of coroutines), but entries without that (i.e. just coroutine function names) also happen

Comment (denesb) [reply scylla-gdb.py:1300]: Ok, we can start with this and refine once we see some broken traces.

Comment (bhalevy) [reply scylla-gdb.py:1300]: I'd consider a regular expression for the right part of the expression above. like
`re.search(r'\) \[clone \.\w+\]$', name)`
(can be optimized by one time re.compile of the regexp above)

**Comment by cmm** [comment]
v2:
- use a compiled regexp to match coro symbols
- put that match in the right place (instead of before every regular match o\)

File: scylla-gdb.py [hunk 2]
```diff
@@ -3644,7 +3644,7 @@ def invoke(self, arg, for_tty):
             initial_task_ptr = int(gdb.parse_and_eval(args.task))
             this_task = self._probe_pointer(initial_task_ptr, args.scanned_region_size, using_seastar_allocator, args.verbose)
             if this_task is None:
-                gdb.write("Provided pointer 0x{:016x} is not an object managed by seastar or not a task pointer\n".format(ptr))
+                gdb.write("Provided pointer 0x{:016x} is not an object managed by seastar or not a task pointer\n".format(initial_task_ptr))
```
Comment (bhalevy) [inline scylla-gdb.py:3647]: nit: if this fix is unrelated to your changes, either submit it in a separate PR
or move it to the front of the series so it would be obvious it is applicable before your changes.

**Review by bhalevy — APPROVED** [review]
Looks good to me

**Comment by cmm** [comment]
v3:
- add a trivial test to check that "task_histogram -a" actually finds some coroutine entries
- reorder commits as suggested

File: scylla-gdb.py [hunk 3]
```diff
@@ -1307,6 +1310,10 @@ def matches_symbol(name):
         return matches_symbol
 
     def __call__(self, name):
+        name = name.strip()
+        if re.search(self._coro_pattern, name) is not None:
```
Comment (bhalevy) [inline scylla-gdb.py:1314]: nit: ` is not None` is not required.
`if re.search(self._coro_pattern, name):` would work just as well, is shorter, and straight forward.

File: test/scylla-gdb/test_misc.py [hunk 4]
```diff
@@ -89,6 +90,11 @@ def test_task_queues(gdb):
 def test_task_histogram(gdb):
     scylla(gdb, 'task_histogram')
 
+def test_task_histogram_coro(gdb):
+    h = scylla(gdb, 'task_histogram -a')
+    if re.search(r'\) \[clone \.\w+\]', h) is None:
```
Comment (bhalevy) [inline test/scylla-gdb/test_misc.py:95]: `if not re.search(r'\) \[clone \.\w+\]', h)` would work just as well
and is more flexible with regards to what is returned (or not) from re.search.

**Review by bhalevy — COMMENTED** [review]
Looks mostly good.
I had only some minor comments on the python code conventions.

File: test/scylla-gdb/test_misc.py (hunk 4 above)
Comment (denesb) [reply test/scylla-gdb/test_misc.py:95]: This flexibility can be undesired. I the canary is `None`, i'd say the current code is better.
I had filtering code break because in addition to the `None` canary, `not filter` also disabled filtering when I filtered for `0`.
